File: other/Yandex - Тренировки по алгоритмам 6.0/4/F.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <utility>

using namespace std;

int main() {
    int N;
    cin >> N;
    
    vector<vector<int>> tree(N + 1);
    for (int i = 0; i < N - 1; i++) {
        int parent;
        cin >> parent;
        tree[parent].push_back(i + 2);
    }
    
    vector<pair<long long, long long>> storage(N + 1);
    
    stack<pair<int, bool>> st;
    st.push({1, false});
    
    while (!st.empty()) {
        auto [node, children_processed] = st.top();
        st.pop();
        
        if (children_processed) {
            if (tree[node].empty()) {
                storage[node] = {1LL, 1LL};
                continue;
            }
            
            long long total_sum = 0, total_count = 0;
            for (int child : tree[node]) {
                auto [child_sum, child_count] = storage[child];
                total_sum += child_sum;
                total_sum += child_count;
                total_count += child_count;
            }
            total_sum += 1;
            total_count += 1;
            storage[node] = {total_sum, total_count};
        } else {
            st.push({node, true});
            for (int child : tree[node]) {
                st.push({child, false});
            }
        }
    }
    
    for (int i = 1; i <= N; i++) {
        cout << storage[i].first << " ";
    }
    cout << endl;
    
    return 0;
}

File: other/Yandex - Тренировки по алгоритмам 6.0/4/F.py
import collections

N = int(input())
storage = [int(el) for el in input().split()]

tree = collections.defaultdict(list)

for i, el in enumerate(storage):
    idx = i + 2
    tree[el].append(idx)

storage = [None] * (N + 1)
stack = [(1, 0)]
while stack:
    node, children_processed = stack.pop()
    if children_processed:
        if not tree[node]:
            storage[node] = (1, 1)
            continue
        total_sum, total_count = (0, 0)
        for child in tree[node]:
            child_sum, child_count = storage[child]
            total_sum += child_sum
            total_sum += child_count
            total_count += child_count
        total_sum += 1
        total_count += 1
        storage[node] = (total_sum, total_count)
    else:
        stack.append((node, 1))
        for child in tree[node]:
            stack.append((child, 0))

for i in range(1, N + 1):
    print(storage[i][0], end=' ')

